pkg/controller/scyllacluster/status.go

```diff
@@ -38,128 +37,60 @@ func (scc *Controller) updateStatus(ctx context.Context, currentSC *scyllav1.Scy
 	return nil
 }
 
-func (scc *Controller) getScyllaVersion(sts *appsv1.StatefulSet) (string, error) {
-	firstMemberName := fmt.Sprintf("%s-0", sts.Name)
-	firstMember, err := scc.podLister.Pods(sts.Namespace).Get(firstMemberName)
-	if err != nil {
-		return "", err
+func (scmc *Controller) calculateStatus(sc *scyllav1.ScyllaCluster, sdcs map[string]*scyllav1alpha1.ScyllaDBDatacenter) scyllav1.ScyllaClusterStatus {
+	status := sc.Status.DeepCopy()
+	if len(sdcs) == 0 {
+		return *status
 	}
 
-	controllerRef := metav1.GetControllerOfNoCopy(firstMember)
-	if controllerRef == nil || controllerRef.UID != sts.UID {
-		return "", fmt.Errorf("foreign pod")
+	sdc, ok := sdcs[sc.Name]
+	if !ok {
+		return *status
 	}
 
-	version, err := naming.ScyllaVersion(firstMember.Spec.Containers)
+	configMaps, err := scmc.configMapLister.ConfigMaps(sdc.Namespace).List(naming.ClusterSelector(sdc))
 	if err != nil {
-		return "", err
-	}
-
-	return version, nil
-}
-
-// calculateRackStatus calculates a status for the rack.
-// sts and old status may be nil.
-func (scc *Controller) calculateRackStatus(sc *scyllav1.ScyllaCluster, rackName string, sts *appsv1.StatefulSet, oldRackStatus *scyllav1.RackStatus, serviceMap map[string]*corev1.Service) *scyllav1.RackStatus {
-	status := &scyllav1.RackStatus{
-		ReplaceAddressFirstBoot: map[string]string{},
-		Members:                 0,
-		ReadyMembers:            0,
-		AvailableMembers:        pointer.Ptr(int32(0)),
-		UpdatedMembers:          pointer.Ptr(int32(0)),
-		Stale:                   pointer.Ptr(true),
-	}
-
-	// Persist ReplaceAddressFirstBoot.
-	if oldRackStatus != nil && oldRackStatus.ReplaceAddressFirstBoot != nil {
-		status.ReplaceAddressFirstBoot = oldRackStatus.DeepCopy().ReplaceAddressFirstBoot
+		klog.ErrorS(err, "can't list configmaps", "Namespace", sdc.Namespace, "Selector", naming.ClusterSelector(sdc))
```

Comment:
why isn't this error returned?

Reply:
This is status, it can't return errors

Reply:
it can, the workque doesn't care who it is

Reply:
Then what you had in mind when you added this comment? It seems right to always suceed for status calculation, and reflect errors within.
https://github.com/scylladb/scylla-operator/blame/master/pkg/controller/scyllacluster/status.go#L145-L147

Reply:
There is a difference between objects missing and a generic List failure - if we have broken listers / clients / REST calls - it should surface an error in the controller logs and also retry.

Effectively, you can always fail before the status calculation when listing / adopting objects. Maybe it would be better if you'd list the objects outside this function and supply it as an argument, similar to controlled objects elsewhere?

Reply:
moved it to caller